src/wallet/rpcdisclosure.cpp

```diff
@@ -0,0 +1,311 @@
+    CDataStream ss(ParseHex(hexInput), SER_NETWORK, PROTOCOL_VERSION);
+    ss >> pd;
+    // if too much data, ok, ignored.
+    // if too little data, --> CBaseDataStream::read(): end of data: iostream error
+
+    // TODO: add errors to an UniValue array and then return array of errors.
+
+    uint256 hash = pd.payload.txid;
+    if (!pwalletMain->mapWallet.count(hash)) {
+        throw JSONRPCError(RPC_MISC_ERROR, "Payment disclosure refers to an invalid or non-wallet transaction id");
+    }
+    const CWalletTx& wtx = pwalletMain->mapWallet[hash];
+
+    UniValue errs(UniValue::VARR);
+    UniValue o(UniValue::VOBJ);
+    o.push_back(Pair("txid", pd.payload.txid.ToString())); // hash instead of txid?
+
+    // validate jsindex against tx
+    if (pd.payload.js >= wtx.vjoinsplit.size()) {
+        errs.push_back("Payment disclosure refers to an invalid joinsplit index");
+        // throw JSONRPCError(RPC_MISC_ERROR, "Payment disclosure refers to an invalid joinsplit index");
+    }
+
+    o.push_back(Pair("jsIndex", pd.payload.js));
+
+    if (pd.payload.n < 0 || pd.payload.n > 1) {
+        errs.push_back("Payment disclosure refers to an invalid output index");
+        // throw JSONRPCError(RPC_MISC_ERROR, "Payment disclosure refers to an invalid output index");
+    }
+    o.push_back(Pair("outputIndex", pd.payload.n));
+
+    if (pd.payload.version < 0 || pd.payload.version > 1) {
+        errs.push_back("Payment disclosure refers to an unknown version");
+        // throw JSONRPCError(RPC_MISC_ERROR, "Payment disclosure refers to an unknown version");
+    }
+    o.push_back(Pair("version", pd.payload.version));
+    o.push_back(Pair("ephemeralSymmetricKey", pd.payload.esk.ToString()));
+    o.push_back(Pair("message", pd.payload.message));
+    // o.push_back(Pair("joinSplitPrivKey", info.joinSplitPrivKey.ToString()));
+    o.push_back(Pair("joinSplitPubKey", wtx.joinSplitPubKey.ToString()));
+
+    if (errs.size() > 0) {
+        o.push_back(Pair("errors", errs));
+    }
+
+    // Sanity check - verify the pd
+    uint256 dataToBeSigned = SerializeHash(pd.payload, SER_GETHASH, 0);
+    bool verified = (crypto_sign_verify_detached(pd.payloadSig.data(),
+        dataToBeSigned.begin(), 32,
+        wtx.joinSplitPubKey.begin()) == 0);
+    o.push_back(Pair("disclosure_sig_verified_same_as_tx", verified));
+    // same private key signed tx as well as this payment disclosure
+    //     throw std::runtime_error("crypto_sign_verify_detached failed");
+    // }
+
+
+
+    // TODO: use esk to decipher the note, to get value (and dest address?)
+
+    JSDescription jsdesc = wtx.vjoinsplit[pd.payload.js];
+
+    uint256 h_sig = jsdesc.h_sig(*pzcashParams, wtx.joinSplitPubKey);
+
+    ZCNoteDecryption decrypter;
+
+    ZCNoteEncryption::Ciphertext ciphertext = jsdesc.ciphertexts[pd.payload.n];
+
+    uint256 pk_enc = zaddr.pk_enc;
+    auto plaintext = decrypter.decryptWithEsk(ciphertext, pk_enc, pd.payload.esk, h_sig, pd.payload.n);
```

Comment:
This doesn't verify the note commitment. See https://github.com/zcash/zips/pull/119/files#r139325414.

Reply:
This has now been fixed in latest update.  Thanks.